internal/controller/apiextensions/claim/configurator.go

```diff
@@ -183,82 +179,94 @@ func filter(in map[string]any, keys ...string) map[string]any {
 	return out
 }
 
-// APIClaimConfigurator configures the supplied claims with fields
-// from the composite. This includes late-initializing spec values
-// and updating status fields in claim.
-type APIClaimConfigurator struct {
-	client client.Client
-}
+func keep(in map[string]any, keys ...string) map[string]any {
+	filter := map[string]bool{}
+	for _, k := range keys {
+		filter[k] = true
+	}
 
-// NewAPIClaimConfigurator returns a APIClaimConfigurator.
-func NewAPIClaimConfigurator(client client.Client) *APIClaimConfigurator {
-	return &APIClaimConfigurator{client: client}
+	out := map[string]any{}
+	for k, v := range in {
+		if filter[k] {
+			out[k] = v
+		}
+	}
+	return out
 }
 
-// Configure the supplied claims with fields from the composite.
+// configure the supplied claim patch with fields from the composite.
 // This includes late-initializing spec values and updating status fields in claim.
-func (c *APIClaimConfigurator) Configure(ctx context.Context, cm resource.CompositeClaim, cp resource.Composite) error {
-	ucm, ok := cm.(*claim.Unstructured)
-	if !ok {
-		return nil
+func configureClaim(_ context.Context, cm *claim.Unstructured, cmPatch *claim.Unstructured, cp *composite.Unstructured, cpPatch *composite.Unstructured) error { //nolint:gocyclo // Only slightly over (10)
+	existing := cm.GetResourceReference()
+	proposed := meta.ReferenceTo(cpPatch, cpPatch.GetObjectKind().GroupVersionKind())
+	equal := cmp.Equal(existing, proposed, cmpopts.IgnoreFields(corev1.ObjectReference{}, "UID"))
+
+	// We refuse to 're-bind' a claim that is already bound to a different
+	// composite resource.
+	if existing != nil && !equal {
+		return errors.New(errBindClaimConflict)
 	}
-	ucp, ok := cp.(*composite.Unstructured)
-	if !ok {
-		return nil
+
+	cmPatch.SetResourceReference(proposed)
+
+	// If existing claim has the status set,
+	// copy the conditions to the patch
+	if s, ok := cm.Object["status"]; ok {
+		fs, ok := s.(map[string]any)
+		if !ok {
+			return errors.Wrap(errors.New(errUnsupportedSrcObject), errMergeClaimStatus)
+		}
+		cmPatch.Object["status"] = keep(fs, "conditions")
+	} else {
+		cmPatch.Object["status"] = map[string]any{}
 	}
 
-	if err := merge(ucm.Object["status"], ucp.Object["status"],
+	// merge from the composite status everything
+	// except conditions and connectionDetails
+	if err := merge(cmPatch.Object["status"], cp.Object["status"],
```

Comment:
Yes, the issue is fixed. I used [the example](https://github.com/turkenh/xp-ssa-experiments/tree/fa71035a49f655802c569481bc7b6038bb89e475/issue-3542) you crafted for the verification, many thanks for that.

For creating the claim patch, we use `status.conditions` from the observed claim and merge all but `status.conditions` from the observed composite. In that way, when a field does not exists in the composite status, it will be removed from claim status as well.

Reply:
I see. 
Couldn't we avoid a merge here? AFAIU, what we need is something like:

```
	cmPatch.Object["status"] = cp.Object["status"]
	cmPatch.Object["status"].(map[string]any)["conditions"] = cp.Object["status"].(map[string]any)["conditions"]
```

\* code is just to give an idea, e.g. leveraging the fieldpath library could help to have a more readable and reliable code.

Reply:
I spend some over this, but I could not come with a more elegant code. The current code does the following:

* keeps `status.conditions` from the observed claim
* merges from the composite everthing except `status.conditions` and `status.connectionDetails`

Therefore, the code looks like this:

```go
// merge from the composite status everything
// except conditions and connectionDetails
if err := merge(cmPatch.Object["status"], cpObserved.Object["status"],
   // Status fields from composite overwrite non-empty fields in claim  
  withMergeOptions(mergo.WithOverride),
  withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceStatusProps())...)); err != nil {
  return errors.Wrap(err, errMergeClaimStatus)
}
```

`merge` functon excepts a filter that gets populated with composite resource status prop names.